lib/rpc/SingleJsonRpcProvider.ts

```diff
@@ -128,29 +139,61 @@ export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
     return super.getBlockNumber()
   }
 
-  private wrappedFunctionCall(fnName: string, fn: (...args: any[]) => Promise<any>, ...args: any[]): Promise<any> {
+  private async wrappedFunctionCall(
+    fnName: string,
+    fn: (...args: any[]) => Promise<any>,
+    ...args: any[]
+  ): Promise<any> {
     this.log.debug(`SingleJsonRpcProvider: wrappedFunctionCall: fnName: ${fnName}, fn: ${fn}, args: ${[...args]}`)
     const perf: SingleCallPerf = {
       succeed: true,
       latencyInMs: 0,
       startTimestampInMs: Date.now(),
     }
-    return fn(...args)
-      .then((response: any) => {
-        return response
-      })
-      .catch((error: any) => {
-        perf.succeed = false
-        this.log.error(JSON.stringify(error))
-        throw error
+    try {
+      return await fn(...args)
+    } catch (error: any) {
+      perf.succeed = false
+      this.log.error(JSON.stringify(error))
+      throw error
+    } finally {
+      perf.latencyInMs = Date.now() - perf.startTimestampInMs
+      this.checkLastCallPerformance(fnName, perf)
+      this.updateHealthyStatus()
+      this.lastCallTimestampInMs = perf.startTimestampInMs
+
+      if (this.enableDbSync) {
+        this.maybeSyncAndUpdateProviderState()
+      }
+    }
+  }
+
+  private maybeSyncAndUpdateProviderState() {
+    this.providerStateSyncer
+      .maybeSyncProviderState(this.healthScore - this.healthScoreAtLastSync, this.healthScore)
+      .then((syncResult: SyncResult) => {
+        if (syncResult.synced) {
+          this.healthScoreAtLastSync = syncResult.state.healthScore
+          this.healthScore = this.healthScoreAtLastSync
+          this.log.debug(`Synced with DB: new health score ${this.healthScore}`)
+          this.updateHealthyStatus()
+        }
       })
-      .finally(() => {
-        perf.latencyInMs = Date.now() - perf.startTimestampInMs
-        this.checkLastCallPerformance(fnName, perf)
-        this.lastCallTimestampInMs = perf.startTimestampInMs
+      .catch((err: any) => {
+        this.log.error(`Encountered unhandled error when sync provider state: ${JSON.stringify(err)}`)
       })
   }
 
+  private updateHealthyStatus() {
+    if (this.healthy && this.healthScore < this.config.HEALTH_SCORE_FALLBACK_THRESHOLD) {
```

Comment:
you could make `this.healthy` a `get` method so you don't need to explicitly set the flag to true/false